### ISSUE TITLE
Infnite tasks support using Jetstream.

### ISSUE THREAD

**Issue by AbstractiveNord** [nats-io/nats-server#4438]
### What motivated this proposal?

I have a dynamic bunch of tasks, which can be stored in Jetstream KV bucket. Every task have some data and should work continuously forever. I am looking for management queue tool for:

1. Load balance tasks between workers.
2. As long as consumer works good (can be verified through ping-pong checks), tasks on him should be treated as running without additional acknowledgement from the consumer.
3. Consumer has to immediately react for task changing. Create or delete should be also treated as change.
4. Every task may have a replicas integer parameter and ability to be filtered by consumers using similar approach as subjects, for avoid same task executing by single consumer.

### What is the proposed change?

New API, similar to jetstream KV bucket, which provides mentioned features.

### Who benefits from this change?

Everyone who wants to use NATS Jetstream queues as task planner for plenty of tasks.

### What alternatives have you evaluated?

NATS Jetstream "IN PROGRESS" acknowledgement, but this signal should be send per every taken message, it causes an unnecessary network load.

**Comment by gedw99** [comment]
https://github.com/choria-io/asyncjobs Does exactly this using nats kv.

but yeah is would be nice if nats Server had this backend in .

cheers

**Comment by ripienaar** [comment]
Personally for this kind of use case I use a combination of KV and WorkQueue. 

KV stores the task state and data while work queue is there to initiate the work that needs to be done and distribute the load out to worker processes etc. 

Workers must tell the server they are doing work and continuing to do so on the “infinite” runtime scenario - nats is not like TCP modeled around connection state - so you have to keep telling it you are still in progress (in progress acks) but you can adjust the frequency at which that is needed using the ack wait. 

this works well for me but ymmv. I am not sure we have plans to add a native abstraction like this similar to KV but if there is an uptick in interest we could

**Comment by AbstractiveNord** [comment]
> Personally for this kind of use case I use a combination of KV and WorkQueue.
> 
> KV stores the task state and data while work queue is there to initiate the work that needs to be done and distribute the load out to worker processes etc.
> 
> Workers must tell the server they are doing work and continuing to do so on the “infinite” runtime scenario - nats is not like TCP modeled around connection state - so you have to keep telling it you are still in progress (in progress acks) but you can adjust the frequency at which that is needed using the ack wait.
> 
> this works well for me but ymmv. I am not sure we have plans to add a native abstraction like this similar to KV but if there is an uptick in interest we could

I've used native stream API for internal KV stream, and your approach works very well, but problem is traffic. Instead of acking with InProgress state every message, borrowed by consumer, it's much better to use a heartbeat or single Ack as an enough acknowledgement, if not any other is send. Can this behavior be emulated with current release of NATS?

**Comment by ripienaar** [comment]
AckInProgress is about as heavy as a heartbeat tbh :) A single ack is enough and thats the default supported behavior. You only need to send in progress when you approach max ack waiting time.

It's a trade off between handling crashed applications (ack timeout till redeliver) small amount of ack traffic, you have to pick a number that works for you.  

How many messages are you talking about where a irregular in progress ack is a problem?

**Comment by AbstractiveNord** [comment]
> AckInProgress is about as heavy as a heartbeat tbh :) A single ack is enough and thats the default supported behavior. You only need to send in progress when you approach max ack waiting time.
> 
> It's a trade off between handling crashed applications (ack timeout till redeliver) small amount of ack traffic, you have to pick a number that works for you.
> 
> How many messages are you talking about where a irregular in progress ack is a problem?

My case is distributed precision time scheduler for tasks, picked up by workers. Schedulers acts like consumers for KV bucket with tasks. Upon task is delivered to scheduler, ack waiting circle is begun. Every scheduler may easily contain 10k-100k+ tasks. Regular acking every message is painful. To make it worse, the lesser ack waiting time, the better for task to be consumed by another scheduler.

**Comment by ripienaar** [comment]
Yes, so how is the consumer supposed to know your worker is still there? You have to tell it.  That's ack in progress.  Its necessary but also optimized for handling many of these so I dont think its a problem

**Comment by AbstractiveNord** [comment]
> Yes, so how is the consumer supposed to know your worker is still there? You have to tell it. That's ack in progress. Its necessary but also optimized for handling many of these so I dont think its a problem

Did I get you correct, it's okay to have a consumer, getting messages in batch mode, and send single InProgress for entire batch?
There is nested structure. Single point of truth is KV bucket with tasks. Each task is some type of command and cron string formatted schedule. Schedulers takes each task, consume it, schedule it, and notify workers via NATS Request-Reply system when it's needed. Issue is all about minimize acks traffic by use consumer heartbit as clear indicator of consumer is active and present.

**Comment by AbstractiveNord** [comment]
> it's okay to have a consumer, getting messages in batch mode, and send single InProgress for entire batch

and have option to nack some messages due to anti-affinity functionality.

**Comment by ripienaar** [comment]
No we dont have batch acks.  You have to send them one by one I am afraid.

You could look to something like a unacknowledged consumer (see ordered consumer) and then check point into a KV bucket your low water mark to resume from later.  Sounds like it might not work too hot for you but thats really the only alternative

**Comment by AbstractiveNord** [comment]
> No we dont have batch acks. You have to send them one by one I am afraid.
> 
> You could look to something like a unacknowledged consumer (see ordered consumer) and then check point into a KV bucket your low water mark to resume from later. Sounds like it might not work too hot for you but thats really the only alternative

It's would be very good to have different ack strategy per type. Some consumer may benefit from nack all batch in one call, or InProgress separate message, or etc. Probably may worth idea about metadated acks. For example, when consumer call acknowledgement, let him optional pass some metadata: should signal affect all messages in batch, or last or first, maybe from begin to current message or from end of batch to current message, maybe several of them, provided via list of indexes, etc. Since NATS is about flexibility, that option may be widely interested by product consumers.

**Comment by ripienaar** [comment]
Do you have multiple clients reading a consumer? Or just one client:one consumer? The All ack strategy only really work with 1:1 client:consumer.

Agree though we do need All versions of in progress and nak its often requested

**Comment by AbstractiveNord** [comment]
> Do you have multiple clients reading a consumer? Or just one client:one consumer? The All ack strategy only really work with 1:1 client:consumer.
> 
> Agree though we do need All versions of in progress and nak its often requested

Yes, I have multiple clients behind single consumer, even dynamically scaling and in different GEOs. In my case of distributed scheduler, every task may have replicas, which is done through multiple consumers on KV task stream.

**Comment by ripienaar** [comment]
Then none of the All acks will work for you.

**Comment by AbstractiveNord** [comment]
> Then none of the All acks will work for you.

Kinda sad, but Issue is all about support this type of communication for NATS.

**Comment by ripienaar** [comment]
When I say it doesnt work I mean the entire concept does not work, its not like oh just add one feature.  It's about the basics of the physical reality we live in that if 10 workers are handling a one if 1 of those worker "ack all" he will ack work for others.

**Comment by AbstractiveNord** [comment]
> When I say it doesnt work I mean the entire concept does not work, its not like oh just add one feature. It's about the basics of the physical reality we live in that if 10 workers are handling work concurrently if 1 of those worker "ack all" he will ack work for others.

Is ACKs about consumer, not about batches?

**Comment by ripienaar** [comment]
AckAll doesnt ack just what you specifically received, its not aware of YOUR outstanding messages.  It just acks all.
